api/lightning/node_cln.py

```diff
@@ -0,0 +1,408 @@
+        """settles a hold invoice"""
+        request = noderpc.HodlInvoiceSettleRequest(payment_hash=hashlib.sha256(bytes.fromhex(preimage)).digest())
+        response = cls.stub.HodlInvoiceSettle(request)
+        
+        return response.state == 2  # True if state is SETTLED, false otherwise.
+
+    @classmethod
+    def gen_hold_invoice(
+        cls, num_satoshis, description, invoice_expiry, cltv_expiry_blocks
+    ):
+        """Generates hold invoice"""
+
+        hold_payment = {}
+        # The preimage is a random hash of 256 bits entropy
+        preimage = hashlib.sha256(secrets.token_bytes(nbytes=32)).digest()
+
+        # Its hash is used to generate the hold invoice
+        r_hash = hashlib.sha256(preimage).digest()
+
+        request = noderpc.InvoiceRequest(
+            description=description,
+            amount_msat=num_satoshis * 1_000,
+            label="TODO", # TODO needs to be a unique string
+            expiry=int(
+                invoice_expiry * 1.5
+            ),  # actual expiry is padded by 50%, if tight, wrong client system clock will say invoice is expired.
+            cltv=cltv_expiry_blocks,
+            preimage=preimage, # preimage is actually optional in cln, as cln would generate one by default
+        )
+        response = cls.stub.HodlInvoice(request)
+
+        hold_payment["invoice"] = response.bolt11
+        payreq_decoded = cls.decode_payreq(hold_payment["invoice"])
+        hold_payment["preimage"] = preimage.hex()
+        hold_payment["payment_hash"] = response.payment_hash
+        hold_payment["created_at"] = timezone.make_aware(
+            datetime.fromtimestamp(payreq_decoded.timestamp)
+        )
+        hold_payment["expires_at"] = response.expires_at
+        hold_payment["cltv_expiry"] = cltv_expiry_blocks
+
+        return hold_payment
+
+    @classmethod
+    def validate_hold_invoice_locked(cls, lnpayment):
+        """Checks if hold invoice is locked"""
+        from api.models import LNPayment
+
+        request = noderpc.HodlInvoiceLookupRequest(
+            payment_hash=bytes.fromhex(lnpayment.payment_hash)
+        )
+        response = cls.stub.HodlInvoiceLookup(request)
+
+        # Will fail if 'unable to locate invoice'. Happens if invoice expiry
+        # time has passed (but these are 15% padded at the moment). Should catch it
+        # and report back that the invoice has expired (better robustness)
+        if response.state == 0:  # OPEN
+            pass
+        if response.state == 1:  # SETTLED
+            pass
+        if response.state == 2:  # CANCELLED
+            pass
+        if response.state == 3:  # ACCEPTED (LOCKED)
+            lnpayment.expiry_height = response.htlc_cltv
+            lnpayment.status = LNPayment.Status.LOCKED
+            lnpayment.save()
+            return True
+
+    @classmethod
+    def resetmc(cls):
+        # don't think an equivalent exists for cln, maybe deleting gossip_store file?
+        return False
```

Comment:
In practice this function is not used. So it is safe to leave it as a blank placeholder.